src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java

```diff
@@ -248,99 +285,146 @@ public Template redactTemplateSecuredFields(User user, Template template) {
+            // Key exists and has been cached successfully
+            listener.onResponse(true);
+            return null;
+        }).exceptionally(throwable -> {
+            Exception exception = SdkClientUtils.unwrapAndConvertToException(throwable);
+            // The cacheMasterKey method only completes exceptionally with FFE
+            if (exception instanceof FlowFrameworkException) {
+                FlowFrameworkException ffe = (FlowFrameworkException) exception;
+                if (ffe.status() == RestStatus.NOT_FOUND) {
+                    // Key doesn't exist, need to generate and index a new one
+                    generateAndIndexNewMasterKey(tenantId, listener);
                 } else {
+                    listener.onFailure(ffe);
+                }
+            } else {
+                // Shouldn't get here
+                listener.onFailure(exception);
+            }
+            return null;
+        });
+    }
+
+    private void generateAndIndexNewMasterKey(String tenantId, ActionListener<Boolean> listener) {
+        String masterKeyId = tenantId == null ? MASTER_KEY : MASTER_KEY + "_" + hashString(tenantId);
+        Config config = new Config(generateMasterKey(), Instant.now());
+        PutDataObjectRequest putRequest = PutDataObjectRequest.builder()
+            .index(CONFIG_INDEX)
+            .id(masterKeyId)
+            .tenantId(tenantId)
+            .dataObject(config)
+            .build();
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+            sdkClient.putDataObjectAsync(putRequest).whenComplete((r, throwable) -> {
+                if (throwable == null) {
                     context.restore();
-                    // Set existing key to master
-                    logger.debug("Config has already been initialized, fetching key");
-                    try (
-                        XContentParser parser = ParseUtils.createXContentParserFromRegistry(
-                            xContentRegistry,
-                            getResponse.getSourceAsBytesRef()
-                        )
-                    ) {
-                        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
-                        Config config = Config.parse(parser);
-                        this.masterKey = config.masterKey();
-                        listener.onResponse(true);
-                    } catch (FlowFrameworkException e) {
-                        listener.onFailure(e);
-                    }
+                    // Set generated key to master
+                    logger.info("Config has been initialized successfully");
+                    setMasterKey(tenantId, config.masterKey());
+                    listener.onResponse(true);
+                } else {
+                    Exception exception = SdkClientUtils.unwrapAndConvertToException(throwable);
```

Comment:
we should also do context.restore() if theres an exception right?

Reply:
Moved it outside the conditional.  It's in a try-with-resources block anyway so is technically unneeded in either.